torchtext/datasets/language_modeling.py

```diff
@@ -1,217 +1,240 @@
-
-        TEXT.build_vocab(train, vectors=vectors)
-
-        return data.BPTTIterator.splits(
-            (train, val, test), batch_size=batch_size, bptt_len=bptt_len,
-            device=device)
 
+        super(LanguageModelingDataset, self).__init__()
+        self._data = data
+        self._vocab = vocab
+
+    def __getitem__(self, i):
+        return self._data[i]
+
+    def __len__(self):
+        return len(self._data)
+
+    def __iter__(self):
+        for x in self._data:
+            yield x
+
+    def get_vocab(self):
+        return self._vocab
+
+
+def _setup_datasets(dataset_name, tokenizer=get_tokenizer("basic_english"),
+                    root='.data', vocab=None, removed_tokens=['<unk>'],
+                    returned_datasets=('train', 'test', 'valid')):
+    train_path = None
+    test_path = None
+    valid_path = None
+    if dataset_name == 'PennTreebank':
+        if 'train' in returned_datasets:
+            train_path = download_from_url(URLS['PennTreebank'][0], root=root)
+        if 'test' in returned_datasets:
+            test_path = download_from_url(URLS['PennTreebank'][1], root=root)
+        if 'valid' in returned_datasets:
+            valid_path = download_from_url(URLS['PennTreebank'][2], root=root)
+    else:
+        dataset_tar = download_from_url(URLS[dataset_name], root=root)
+        extracted_files = extract_archive(dataset_tar)
+
+        for fname in extracted_files:
+            if 'train' in returned_datasets and 'train' in fname:
+                train_path = os.path.join(root, fname)
+            elif 'test' in returned_datasets and 'test' in fname:
+                test_path = os.path.join(root, fname)
+            elif 'valid' in returned_datasets and 'valid' in fname:
+                valid_path = os.path.join(root, fname)
+
+    if vocab is None:
+        logging.info('Building Vocab based on {}'.format(train_path))
+        if train_path is None:
+            raise TypeError("Train file is not defined correctly to generate vocabulary")
+        vocab = build_vocab_from_iterator(read_text_iterator(train_path, tokenizer))
+        logging.info('Vocab has {} entries'.format(len(vocab)))
+    else:
+        if not isinstance(vocab, Vocab):
+            raise TypeError("Passed vocabulary is not of type Vocab")
+
+    train_data = []
+    if train_path is not None:
+        logging.info('Creating training data')
+        train_iter = create_data_from_iterator(
+            vocab, read_text_iterator(train_path, tokenizer), removed_tokens)
+        for tokens in train_iter:
+            train_data += tokens
+
+    test_data = []
+    if test_path is not None:
+        logging.info('Creating testing data')
+        test_iter = create_data_from_iterator(
+            vocab, read_text_iterator(test_path, tokenizer), removed_tokens)
+        for tokens in test_iter:
+            test_data += tokens
+
+    valid_data = []
+    if valid_path is not None:
```

Comment:
I think you could save a bit of space by writing something like

```
    paths = {}
    for fname in extracted_files:
        for k in ['train', 'valid', 'test']:
            if k in fname:
                paths[k] = os.path.join(root, fname)

    def gen_data(path_key, paths):
        path = paths[path_key]
        logging.info('Creating {} data'.format(path_key))
        return list(tokens for tokens in create_data_from_iterator(
            vocab, read_text_iterator(test_path, tokenizer), removed_tokens))

    return tuple(LanguageModelingDataset(torch.tensor(gen_data(k), vocab) for k in returned_datasets)
```

List expressions also tend to be faster than explicit for-loops.

Reply:
Fixed!